caas/kubernetes/provider/application/application_test.go

```diff
@@ -2988,6 +3041,347 @@ func (s *applicationSuite) TestEnsureUpdatedConstraints(c *gc.C) {
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "sa1",
+			Namespace: modelName,
+			Labels:    resourceLabels,
+		},
+	}
+	_, err := s.client.CoreV1().ServiceAccounts(modelName).Create(ctx, sa, metav1.CreateOptions{FieldManager: "juju"})
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Role (namespace-scoped)
+	role := &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "role1",
+			Namespace: modelName,
+			Labels:    resourceLabels,
+		},
+		Rules: []rbacv1.PolicyRule{{
+			APIGroups: []string{""}, Resources: []string{"pods"}, Verbs: []string{"get", "list"},
+		}},
+	}
+	_, err = s.client.RbacV1().Roles(modelName).Create(ctx, role, metav1.CreateOptions{FieldManager: "juju"})
+	c.Assert(err, jc.ErrorIsNil)
+
+	// RoleBinding (namespace-scoped)
+	rb := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rb1",
+			Namespace: modelName,
+			Labels:    resourceLabels,
+		},
+		RoleRef:  rbacv1.RoleRef{APIGroup: "rbac.authorization.k8s.io", Kind: "Role", Name: role.Name},
+		Subjects: []rbacv1.Subject{{Kind: "ServiceAccount", Name: sa.Name, Namespace: modelName}},
+	}
+	_, err = s.client.RbacV1().RoleBindings(modelName).Create(ctx, rb, metav1.CreateOptions{FieldManager: "juju"})
+	c.Assert(err, jc.ErrorIsNil)
+
+	// ClusterRole (cluster-scoped)
+	cr := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   "cr1",
+			Labels: resourceLabels,
+		},
+		Rules: []rbacv1.PolicyRule{{
+			APIGroups: []string{""}, Resources: []string{"nodes"}, Verbs: []string{"get", "list"},
+		}},
+	}
+	_, err = s.client.RbacV1().ClusterRoles().Create(ctx, cr, metav1.CreateOptions{FieldManager: "juju"})
+	c.Assert(err, jc.ErrorIsNil)
+
+	// ClusterRoleBinding (cluster-scoped)
+	crb := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   "crb1",
+			Labels: resourceLabels,
+		},
+		RoleRef:  rbacv1.RoleRef{APIGroup: "rbac.authorization.k8s.io", Kind: "ClusterRole", Name: cr.Name},
+		Subjects: []rbacv1.Subject{{Kind: "ServiceAccount", Name: sa.Name, Namespace: modelName}},
+	}
+	_, err = s.client.RbacV1().ClusterRoleBindings().Create(ctx, crb, metav1.CreateOptions{FieldManager: "juju"})
+	c.Assert(err, jc.ErrorIsNil)
+
+	// ConfigMap
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "cm1",
+			Namespace: modelName,
+			Labels:    resourceLabels,
+		},
+		Data: map[string]string{"k": "v"},
+	}
+	_, err = s.client.CoreV1().ConfigMaps(modelName).Create(ctx, cm, metav1.CreateOptions{FieldManager: "juju"})
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Secret
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-container-secret", appName),
+			Namespace: modelName,
+			Labels:    resourceLabels,
+		},
+		Type: corev1.SecretTypeDockerConfigJson,
+		Data: map[string][]byte{corev1.DockerConfigJsonKey: []byte(`{"auths":{}}`)},
+	}
+	_, err = s.client.CoreV1().Secrets(modelName).Create(ctx, secret, metav1.CreateOptions{FieldManager: "juju"})
```

Comment:
For each of these resources, we should create 2 - one to be deleted and one to be left behind and check.